proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,161 @@
+| kroxylicious_upstream_errors                     | Counter      | flowing, virtualCluster                     | Incremented by one each time a connection attempt fails owning to a upstream error (tls negotiation etc)            |
+
+### Specific short-comings/weaknesses
+
+* *1 - these metrics record only the **requests**, there's nothing analogous that records the responses.
+* *1 - these metrics aren't discriminated by ApiKey (etc.). This limits their usefulness for problem-solving.
+* *2 - the metric name is quite misleading - nothing tells the user that it is only recording the sizes of RPCs that are actually **decoded**.
+* *2 - there are no quantised buckets configured for this metric. it gives the user count/sum/max values only. it would be
+  hard to use this metric to understand something about produce requests or fetch responses.
+* *1 / *2 - we have contradictory meanings for the `flowing` label.  in *2, requests flow `UPSTREAM`, responses flow `DOWNSTREAM`.
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+
+## Use Cases
+
+In general, metrics exist to help a user understand the health and performance of the proxy. They help inform
+capacity planning, drive system alerts reporting abnormal conditions and help with problem-solving (things like
+diagnosing failures or system slow-downs).
+
+Introducing a proxy to a Kafka system will increase latency of the system. Users will need a mechanism to understand how
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+
+## Proposal
+
+### Deprecate `kroxylicious_inbound*`/`kroxylicious_payload_size_bytes` replace with new `kroxylicious_(downstream|upstream)_kafka_rpc_message_size_bytes` metrics
+
+This proposal suggests that we deprecate the `kroxylicious_inbound*`/`kroxylicious_payload_size_bytes` metrics and replace
+with two new **distribution** (aka histogram) metrics `kroxylicious_downstream_kafka_rpc_message_size_bytes` and `kroxylicious_upstream_kafka_rpc_message_size_bytes`.
```

Comment:
aren't `kafka_rpc` and `message` synonyms in this context? Does it help users to repeat them?

Reply:
Yes, you are right.  The Kafka protocol documentation uses the word message.  I'll use the term kafka_message.

Reply:
Fixed.